keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2304 @@
+// Drain:
+// 1. Cordons all selected nodes by making them unschedulable.
+// 2. Gives an intent to drain all selected nodes by creating EvictionRequest objects for the
+//    node's pods.
+Drain NodeMaintenanceStage = "Drain"
+// Complete:
+// 1. Removes all EvictionRequest objects requested by this NodeMaintenance.
+// 2. Uncordons all selected nodes by making them schedulable again, unless there is not another
+//    maintenance in progress.
+Complete NodeMaintenanceStage = "Complete"
+)
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    // NodeSelector selects nodes for this node maintenance.
+    // +required
+    NodeSelector *v1.NodeSelector
+
+    // The order of the stages is Idle -> Cordon -> Drain -> Complete.
+    //
+    // - The Cordon or Drain stage can be skipped by setting the stage to Complete.
+    // - The NodeMaintenance object is moved to the Complete stage on deletion unless the Idle stage has been set.
+    //
+    // The default value is Idle.
+    Stage NodeMaintenanceStage
+
+    // DrainPlan is executed from the first entry to the last entry during the Drain stage.
+    // DrainPlanEntry podType fields should be in the following order:
+    // nil -> DaemonSet -> Static
+    // DrainPlanEntry priority fields should be in ascending order for each podType.
+    // If the priority and podType are the same, concrete selectors are executed first.
+    //
+    // The following entries are injected into the drainPlan on the NodeMaintenance admission:
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "Default"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "Default"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "Default"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "Default"
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "DaemonSet"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "DaemonSet"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "DaemonSet"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "DaemonSet"
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "Static"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "Static"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "Static"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "Static"
+    //
+    // Duplicate entries are not allowed.
+    // This field is immutable.
+    DrainPlan []DrainPlanEntry
+
+    // Reason for the maintenance.
+    Reason string
```

Comment:
Will it be enough to use `string` here to cover all scenarios? 
For some cases it will be better to have smth "machine readable" here.

User story:
As a platform engineer I want to provide machine readable reason for the users of the platform. To empower user to write some automations based on this reason.  
For that case better to write some abstract interface here. For example - to use ObjectReference here like cluster-api do with infra-provider.